File: naiveBayesClassifier/src/test/java/org/metplus/curriculum/cruncher/naivebayes/CruncherImplTest.java
package org.metplus.curriculum.cruncher.naivebayes;

import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Suite;
import org.metplus.curriculum.test.BeforeAfterInterface;
import org.metplus.curriculum.test.BeforeAfterRule;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import static org.junit.Assert.assertEquals;

/**
 * Created by joao on 9/8/16.
 */
@RunWith(Suite.class)
@Suite.SuiteClasses({CruncherImplTest.TrainWithMap.class,
        CruncherImplTest.TrainSingleFeature.class})
public class CruncherImplTest {
    public static class Base implements BeforeAfterInterface {
        @Rule
        public BeforeAfterRule beforeAfter = new BeforeAfterRule(this);
        protected CruncherImpl cruncher;
        @Override
        public void after() {

        }

        @Override
        public void before() {
            cruncher = new CruncherImpl();
        }
    }

    public static class TrainWithMap extends Base {
        @Test
        public void nullSet() {
            cruncher.train(null);
            assertEquals(0, cruncher.getClassifier().getCategoriesTotal());
        }
        @Test
        public void emptySet() {
            cruncher.train(new HashMap<>());
            assertEquals(0, cruncher.getClassifier().getCategories().size());
        }

        @Test
        public void oneFeature() {
            HashMap<String, List<String>> database = new HashMap<>();
            ArrayList<String> feature = new ArrayList<>();
            feature.add("nice");
            feature.add("good");
            feature.add("weather");

            database.put("bamm", feature);
            cruncher.train(database);
            assertEquals(1, cruncher.getClassifier().getCategories().size());
        }

        @Test
        public void twoFeatures() {
            HashMap<String, List<String>> database = new HashMap<>();
            ArrayList<String> feature = new ArrayList<>();
            feature.add("nice");
            feature.add("good");
            feature.add("weather");

            database.put("bamm", feature);
            database.put("bamm1", feature);
            cruncher.train(database);
            assertEquals(2, cruncher.getClassifier().getCategories().size());
        }
    }

    public static class TrainSingleFeature extends Base {
        @Test
        public void nullSet() {
            cruncher.train("bamm", null);
            assertEquals(0, cruncher.getClassifier().getCategoriesTotal());
        }
        @Test
        public void emptySet() {
            cruncher.train("bamm", new ArrayList<>());
            assertEquals(0, cruncher.getClassifier().getCategoriesTotal());
        }
        @Test
        public void oneFeature() {
            ArrayList<String> feature = new ArrayList<>();
            feature.add("nice");
            feature.add("good");
            feature.add("weather");

            cruncher.train("bamm", feature);
            assertEquals(1, cruncher.getClassifier().getCategories().size());
        }
    }
}

File: crunchers/naiveBayes/src/test/kotlin/org/metplus/curriculum/cruncher/naivebayes/CruncherImplTest.kt
package org.metplus.curriculum.cruncher.naivebayes


import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.util.*

class CruncherImplTest {
    class TrainWithMap {
        private lateinit var cruncher: CruncherImpl

        @BeforeEach
        fun before() {
            this.cruncher = CruncherImpl()
        }

        @Test
        fun emptySet() {
            cruncher.train(emptyMap())
            assertThat(cruncher.classifier.categories.size).isEqualTo(0)
        }

        @Test
        fun oneFeature() {
            val database = HashMap<String, List<String>>()
            val feature = ArrayList<String>()
            feature.add("nice")
            feature.add("good")
            feature.add("weather")

            database["bamm"] = feature
            cruncher.train(database)
            assertThat(cruncher.classifier.categories.size).isEqualTo(1)
        }

        @Test
        fun twoFeatures() {
            val database = HashMap<String, List<String>>()
            val feature = ArrayList<String>()
            feature.add("nice")
            feature.add("good")
            feature.add("weather")

            database["bamm"] = feature
            database["bamm1"] = feature
            cruncher.train(database)
            assertThat(cruncher.classifier.categories.size).isEqualTo(2)
        }
    }

    class TrainSingleFeature {
        private lateinit var cruncher: CruncherImpl

        @BeforeEach
        fun before() {
            this.cruncher = CruncherImpl()
        }

        @Test
        fun nullSet() {
            cruncher.train("bamm", null)
            assertThat(cruncher.classifier.categoriesTotal).isEqualTo(0)
        }

        @Test
        fun emptySet() {
            cruncher.train("bamm", emptyList())
            assertThat(cruncher.classifier.categoriesTotal).isEqualTo(0)
        }

        @Test
        fun oneFeature() {
            val feature = ArrayList<String>()
            feature.add("nice")
            feature.add("good")
            feature.add("weather")

            cruncher.train("bamm", feature)
            assertThat(cruncher.classifier.categories.size).isEqualTo(1)
        }
    }

    class Crunch {
        private lateinit var cruncher: CruncherImpl

        @BeforeEach
        fun before() {
            this.cruncher = CruncherImpl()
            cruncher.train(mapOf(
                    "positive" to listOf("I love sunny days"),
                    "negative" to listOf("I hate rain")
            ))
        }

        @Test
        fun checkCruncherResults() {
            val unknownText1 = "today is a sunny day"
            val unknownText2 = "there will be rain"

            assertThat(cruncher.crunch(unknownText1).metaData["positive"]).isGreaterThan(cruncher.crunch(unknownText1).metaData["negative"])
            assertThat(cruncher.crunch(unknownText2).metaData["negative"]).isGreaterThan(cruncher.crunch(unknownText2).metaData["positive"])
        }
    }
}
